# dedup benchmarking of GBS ft. Bees

---

## Example Datasets

- bclconvert PATH: /bifo/scratch/hiseq/postprocessing/illumina/novaseq/$RUNROOT/SampleSheet/bclconvert; these reads are not dereplicated.
 
- link_farm PATH: /dataset/hiseq/active/fastq-link-farm; these reads are dereplicated.

```
bclconvert                      link_farm                          cohort_moniker                  phenotype
SQ1766_S3_L001_R1_001.fastq.gz  SQ1766_H322HDMXY_s_1_fastq.txt.gz  SQ1766.all.deer.PstI            slippers
SQ1766_S3_L002_R1_001.fastq.gz  SQ1766_H322HDMXY_s_2_fastq.txt.gz  SQ1766.all.deer.PstI            slippers

SQ1769_S6_L001_R1_001.fastq.gz  SQ1769_H322HDMXY_s_1_fastq.txt.gz  SQ1769.all.deer.PstI            noslips
SQ1769_S6_L002_R1_001.fastq.gz  SQ1769_H322HDMXY_s_2_fastq.txt.gz  SQ1769.all.deer.PstI            noslips

SQ1793_S2_L001_R1_001.fastq.gz  SQ1793_HN7WGDRXY_s_1_fastq.txt.gz  SQ1793.all.bee.MspI-ApeKI       noslips
SQ1793_S2_L002_R1_001.fastq.gz  SQ1793_HN7WGDRXY_s_2_fastq.txt.gz  SQ1793.all.bee.MspI-ApeKI       noslips

SQ1794_S3_L001_R1_001.fastq.gz  SQ1794_HN7WGDRXY_s_1_fastq.txt.gz  SQ1794.all.bee.MspI-ApeKI       noslips
SQ1794_S3_L002_R1_001.fastq.gz  SQ1794_HN7WGDRXY_s_2_fastq.txt.gz  SQ1794.all.bee.MspI-ApeKI       noslips

SQ1795_S1_L001_R1_001.fastq.gz  SQ1795_HN7GKDRXY_s_1_fastq.txt.gz  SQ1795.all.bee_not_kauri.ApeKI  noslips likely sub-set of SQ1794 (resequenced)
SQ1795_S1_L002_R1_001.fastq.gz  SQ1795_HN7GKDRXY_s_1_fastq.txt.gz  SQ1795.all.bee_not_kauri.ApeKI  noslips likely sub-set of SQ1794 (resequenced)


```
## gquery examples for related keyfile data
```
gquery
-t gbs_keyfile
-b taxname
-p "columns=flowcell,libraryprepid,enzyme;sequencing_platform=novaseq;no_unpivot;group"
bee
```
```
flowcell        libraryprepid   enzyme          n
HHNNVDRXY       SQ1647          MspI-ApeKI      175
HN7WGDRXY       SQ1793          MspI-ApeKI      80
HN7WGDRXY       SQ1794          MspI-ApeKI      77
```
---
```
gquery
-t gbs_keyfile
-b taxname
-p "columns=flowcell,libraryprepid,enzyme;sequencing_platform=novaseq;no_unpivot;group"
kauri
```
```
flowcell        libraryprepid   enzyme          n
HN7GKDRXY       1795            ApeKI           74  - this is actually Bees
HN7GKDRXY       1796            PstI-MspI       74
HN7GKDRXY       1806            SbfI-MspI       74
```
---

## Command line gbs qc calls

Examples of restart_qc.src scripts generated by ./run_gbs_qc interactive mode.
```
export SEQ_PRISMS_BIN=/dataset/gseq_processing/active/bin/gbs_prism/seq_prisms
export GBS_PRISM_BIN=/dataset/gseq_processing/active/bin/gbs_prism

/dataset/gseq_processing/active/bin/gbs_prism/ag_gbs_qc_prism.sh \
-m novaseq \
-C slurm \
-p /dataset/gseq_processing/active/bin/gbs_prism/nomerge_options_file.txt \
-a unblind \
-O /dataset/hiseq/scratch/postprocessing/gbs/220322_A01439_0058_AHN7GKDRXY \
-r 220322_A01439_0058_AHN7GKDRXY \
# libraries
SQ1795.all.bee_not_kauri.ApeKI
```
---
```
export SEQ_PRISMS_BIN=/dataset/gseq_processing/active/bin/gbs_prism/seq_prisms
export GBS_PRISM_BIN=/dataset/gseq_processing/active/bin/gbs_prism

/dataset/gseq_processing/active/bin/gbs_prism/ag_gbs_qc_prism.sh 
-m novaseq 
-C slurm 
-p /dataset/gseq_processing/active/bin/gbs_prism/nomerge_options_file.txt 
-a unblind 
-O /dataset/hiseq/scratch/postprocessing/gbs/220322_A01439_0058_AHN7GKDRXY 
-r 220317_A01439_0057_BHN7WGDRXY 
# libraries
SQ1793.all.bee.MspI-ApeKI 
SQ1794.all.bee.MspI-ApeKI
```
---

# Dereplication Software
1. vsearch
```
vsearch \
--fastx_uniques seqs.fastq.gz \
--threads 16 \
--gzip \
--minuniquesize 1 \
--fastqout uniques.fastq.gz \
--uc vsearch-derep.uc \
--relabel_sha1 \
--relabel_keep \
--sizeout \
--tabbedout vsearch.derep.summary.txt
```
2. BBMap, current method, uses cluster proximity on the flowcell to constrain dereplication
```
/usr/bin/time \
clumpify.sh \
dedupe \
optical \
dupedist=15000 \
subs=0 \
tmpdir=/dataset/gseq_processing/itmp/illumina/clumpifytAtSR \
in=/dataset/hiseq/scratch/postprocessing/illumina/novaseq/220225_A01439_0052_BH322HDMXY/SampleSheet/bclconvert/SQ1764_S1_L001_R1_001.fastq.gz \
out=/dataset/hiseq/scratch/postprocessing/illumina/novaseq/220225_A01439_0052_BH322HDMXY/SampleSheet/dedup/15000/SQ1764_S1_L001_R1_001.fastq.gz \ 
2>/dataset/hiseq/scratch/postprocessing/illumina/novaseq/220225_A01439_0052_BH322HDMXY/SampleSheet/dedup/15000/SQ1764_S1_L001_R1_001.fastq.gz.stderr \
1>/dataset/hiseq/scratch/postprocessing/illumina/novaseq/220225_A01439_0052_BH322HDMXY/SampleSheet/dedup/15000/SQ1764_S1_L001_R1_001.fastq.gz.stdout
```
3. usearch, requires 64-bit license to be viable (which we do not currently have). Current license is cor v7 which does not support fastq dereplication or denoising.
```
usearch -unoise3 uniques.fa -zotus otus100.fa
```
then,
```
for id in 99 97 95 90;
do 
   usearch -cluster_smallmem otus100.fa -id 0.$id -centroids otus$id.fa;
done;
```
4. dada2 ASV calculation - not ideal as requires R script.
    Example dada2 script,
```
library(dada2); packageVersion("dada2")

# File parsing
filtpath <- "path/to/FWD/filtered" # CHANGE ME to the directory containing your filtered fastq files
filts <- list.files(filtpath, pattern="fastq.gz", full.names=TRUE) # CHANGE if different file extensions
sample.names <- sapply(strsplit(basename(filts), "_"), `[`, 1) # Assumes filename = sample_XXX.fastq.gz
names(filts) <- sample.names

# Learn error rates
set.seed(100)
err <- learnErrors(filts, nbases = 1e8, multithread=TRUE, randomize=TRUE)
plotErrors(errF, nominalQ=TRUE)
   
# Infer sequence variants
dds <- vector("list", length(sample.names))
names(dds) <- sample.names

for(sam in sample.names) {
   cat("Processing:", sam, "\n")
   derep <- derepFastq(filts[[sam]])
   dds[[sam]] <- dada(derep, err=err, multithread=TRUE)
}

# Construct sequence table and write to disk
seqtab <- makeSequenceTable(dds)
sq <- getSequences(seqtab)
id <- paste0("Abundance=", colSums(seqtab))
names(sq) <- id

writeFasta(sq, file="path/to/myasvs.fasta")
   
```