# dedup benchmarking of GBS ft. Bees

---
## gquery examples for related keyfile data
```
gquery
-t gbs_keyfile
-b taxname
-p "columns=flowcell,libraryprepid,enzyme;sequencing_platform=novaseq;no_unpivot;group"
bee
```
```
flowcell        libraryprepid   enzyme          n
HHNNVDRXY       SQ1647          MspI-ApeKI      175
HN7WGDRXY       SQ1793          MspI-ApeKI      80
HN7WGDRXY       SQ1794          MspI-ApeKI      77
```
---
```
gquery
-t gbs_keyfile
-b taxname
-p "columns=flowcell,libraryprepid,enzyme;sequencing_platform=novaseq;no_unpivot;group"
kauri
```
```
flowcell        libraryprepid   enzyme          n
HN7GKDRXY       1795            ApeKI           74  - this is actually Bees
HN7GKDRXY       1796            PstI-MspI       74
HN7GKDRXY       1806            SbfI-MspI       74
```
## Command line gbs qc calls

Examples of restart_qc.src scripts generated by ./run_gbs_qc interactive mode.
```
export SEQ_PRISMS_BIN=/dataset/gseq_processing/active/bin/gbs_prism/seq_prisms
export GBS_PRISM_BIN=/dataset/gseq_processing/active/bin/gbs_prism

/dataset/gseq_processing/active/bin/gbs_prism/ag_gbs_qc_prism.sh \
-m novaseq \
-C slurm \
-p /dataset/gseq_processing/active/bin/gbs_prism/nomerge_options_file.txt \
-a unblind \
-O /dataset/hiseq/scratch/postprocessing/gbs/220322_A01439_0058_AHN7GKDRXY \
-r 220322_A01439_0058_AHN7GKDRXY \
# libraries
SQ1795.all.bee_not_kauri.ApeKI
```
---
```
export SEQ_PRISMS_BIN=/dataset/gseq_processing/active/bin/gbs_prism/seq_prisms
export GBS_PRISM_BIN=/dataset/gseq_processing/active/bin/gbs_prism

/dataset/gseq_processing/active/bin/gbs_prism/ag_gbs_qc_prism.sh 
-m novaseq 
-C slurm 
-p /dataset/gseq_processing/active/bin/gbs_prism/nomerge_options_file.txt 
-a unblind 
-O /dataset/hiseq/scratch/postprocessing/gbs/220322_A01439_0058_AHN7GKDRXY 
-r 220317_A01439_0057_BHN7WGDRXY 
# libraries
SQ1793.all.bee.MspI-ApeKI 
SQ1794.all.bee.MspI-ApeKI
```

# Dereplication Software

1. `vsearch --fastx_uniques seqs --output rep-set.fna --uc vsearch-derep.uc --relabel_sha1 --relabel_keep`


2. `usearch -unoise3 uniques.fa -zotus otus100.fa`
    then,
    ```
   for id in 99 97 95 90
   do
   usearch -cluster_smallmem otus100.fa -id 0.$id -centroids otus$id.fa
   done
   ```
   
3. dada2 ASV calculation - not ideal as requires R script.
    Example dada2 script,
    ```
    library(dada2); packageVersion("dada2")
    
    # File parsing
    filtpath <- "path/to/FWD/filtered" # CHANGE ME to the directory containing your filtered fastq files
    filts <- list.files(filtpath, pattern="fastq.gz", full.names=TRUE) # CHANGE if different file extensions
    sample.names <- sapply(strsplit(basename(filts), "_"), `[`, 1) # Assumes filename = sample_XXX.fastq.gz
    names(filts) <- sample.names
    
    # Learn error rates
    set.seed(100)
    err <- learnErrors(filts, nbases = 1e8, multithread=TRUE, randomize=TRUE)
    plotErrors(errF, nominalQ=TRUE)
   
    # Infer sequence variants
    dds <- vector("list", length(sample.names))
    names(dds) <- sample.names
   
    for(sam in sample.names) {
       cat("Processing:", sam, "\n")
       derep <- derepFastq(filts[[sam]])
       dds[[sam]] <- dada(derep, err=err, multithread=TRUE)
    }
    
    # Construct sequence table and write to disk
    seqtab <- makeSequenceTable(dds)
    sq <- getSequences(seqtab)
    id <- paste0("Abundance=", colSums(seqtab))
    names(sq) <- id
    
    writeFasta(sq, file="path/to/myasvs.fasta")
   
   ```